File: lib/rds.ts
import * as aws from '@pulumi/aws';
import {
  commonTags,
  dbInstanceClass,
  primaryRegion,
  secondaryRegion,
} from './config';
import { primaryKmsKey, secondaryKmsKey } from './kms';
import {
  primaryDbSecurityGroup,
  secondaryDbSecurityGroup,
} from './security-groups';
import {
  primaryPrivateSubnet1,
  primaryPrivateSubnet2,
  secondaryPrivateSubnet1,
  secondaryPrivateSubnet2,
} from './vpc';

const primaryProvider = new aws.Provider('primary-provider', {
  region: primaryRegion,
});
const secondaryProvider = new aws.Provider('secondary-provider', {
  region: secondaryRegion,
});

// Primary region DB subnet group
export const primaryDbSubnetGroup = new aws.rds.SubnetGroup(
  'primary-db-subnet-group',
  {
    name: 'primary-db-subnet-group',
    subnetIds: [primaryPrivateSubnet1.id, primaryPrivateSubnet2.id],
    tags: {
      ...commonTags,
      Name: 'Primary DB Subnet Group',
    },
  },
  { provider: primaryProvider }
);

// Secondary region DB subnet group
export const secondaryDbSubnetGroup = new aws.rds.SubnetGroup(
  'secondary-db-subnet-group',
  {
    name: 'secondary-db-subnet-group',
    subnetIds: [secondaryPrivateSubnet1.id, secondaryPrivateSubnet2.id],
    tags: {
      ...commonTags,
      Name: 'Secondary DB Subnet Group',
    },
  },
  { provider: secondaryProvider }
);

// Primary RDS instance
export const primaryRdsInstance = new aws.rds.Instance(
  'primary-mysql-db',
  {
    identifier: 'primary-mysql-database',
    allocatedStorage: 20,
    storageType: 'gp2',
    storageEncrypted: true,
    kmsKeyId: primaryKmsKey.arn,
    engine: 'mysql',
    engineVersion: '5.7',
    instanceClass: dbInstanceClass,
    dbName: 'productiondb',
    username: 'admin',
    password: 'changeme123!', // In production, use AWS Secrets Manager
    vpcSecurityGroupIds: [primaryDbSecurityGroup.id],
    dbSubnetGroupName: primaryDbSubnetGroup.name,
    backupRetentionPeriod: 7,
    backupWindow: '03:00-04:00',
    maintenanceWindow: 'sun:04:00-sun:05:00',
    skipFinalSnapshot: true,
    deletionProtection: false, // Set to true in production
    tags: {
      ...commonTags,
      Name: 'Primary MySQL Database',
    },
  },
  { provider: primaryProvider }
);

// Cross-region read replica
export const secondaryRdsReadReplica = new aws.rds.Instance(
  'secondary-mysql-read-replica',
  {
    identifier: 'secondary-mysql-read-replica',
    replicateSourceDb: primaryRdsInstance.arn,
    instanceClass: dbInstanceClass,
    storageEncrypted: true,
    kmsKeyId: secondaryKmsKey.arn,
    vpcSecurityGroupIds: [secondaryDbSecurityGroup.id],
    dbSubnetGroupName: secondaryDbSubnetGroup.name,
    skipFinalSnapshot: true,
    deletionProtection: false, // Set to true in production
    tags: {
      ...commonTags,
      Name: 'Secondary MySQL Read Replica',
    },
  },
  { provider: secondaryProvider }
);

File: archive/pulumi-py/Pr6231/lib/rds.py
"""
RDS PostgreSQL Database
Creates RDS instance with automated backups
"""

import pulumi
import pulumi_aws as aws
import json

def create_rds_instance(environment_suffix: str, vpc, private_subnets, security_group):
    """
    Create RDS PostgreSQL instance with automated backups
    """

    # Create DB subnet group
    db_subnet_group = aws.rds.SubnetGroup(
        f"db-subnet-group-{environment_suffix}",
        subnet_ids=[subnet.id for subnet in private_subnets],
        tags={
            "Name": f"db-subnet-group-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    # Create a random password for the database
    db_password = pulumi.Output.secret("TempPassword123!")  # In production, use existing secret
    db_username = "postgres"
    db_name = "flaskapp"

    # Create RDS PostgreSQL instance
    db_instance = aws.rds.Instance(
        f"postgres-{environment_suffix}",
        identifier=f"postgres-{environment_suffix}",
        engine="postgres",
        engine_version="14",
        instance_class="db.t3.micro",
        allocated_storage=20,
        storage_type="gp2",
        storage_encrypted=True,
        db_name=db_name,
        username=db_username,
        password=db_password,
        db_subnet_group_name=db_subnet_group.name,
        vpc_security_group_ids=[security_group.id],
        publicly_accessible=False,
        skip_final_snapshot=True,  # For destroyable infrastructure
        backup_retention_period=5,
        backup_window="03:00-04:00",
        maintenance_window="mon:04:00-mon:05:00",
        multi_az=False,  # Single AZ for cost optimization
        tags={
            "Name": f"postgres-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    # Create connection string secret in Secrets Manager
    db_connection_string = pulumi.Output.all(
        db_instance.endpoint,
        db_username,
        db_password,
        db_name
    ).apply(lambda args: f"postgresql://{args[1]}:{args[2]}@{args[0]}/{args[3]}")

    db_secret = aws.secretsmanager.Secret(
        f"db-connection-{environment_suffix}",
        name=f"db-connection-{environment_suffix}",
        description="Database connection string for Flask application",
        tags={
            "Name": f"db-connection-{environment_suffix}",
            "EnvironmentSuffix": environment_suffix
        }
    )

    db_secret_version = aws.secretsmanager.SecretVersion(
        f"db-connection-version-{environment_suffix}",
        secret_id=db_secret.id,
        secret_string=db_connection_string
    )

    return {
        "db_instance": db_instance,
        "db_subnet_group": db_subnet_group,
        "db_secret": db_secret
    }
